### PR TITLE
Training the `while_let` detector...

### LINKED ISSUE #262: Extend `while_let_loop` to warn even if there are statements after `match`
The implementation of `while_let_loop` is incomplete. In cases when additional statements follow the `match`, such as the following:
```
loop {
    let bi = match bodies.next() {
        Some(bi) => bi,
        None => break
    };
    additional_statement1(bi);
    additional_statement2();
}
```
the loop can still be replaced with `while let`
```
while let Some(bi) = bodies.next() {
    additional_statement1(bi);
    additional_statement2();
}
```
These snippets are slightly different in theory, but identical in practice.

### PR DESCRIPTION
**Pull request by Wafflespeanut**
fixes #262

### CODE REVIEW

File: tests/compile-fail/while_loop.rs [hunk 1]
```diff
@@ -1,20 +1,12 @@
...
             let _v = 1;
         } else {
-            break;
-        }
-    }
-    loop { //~ERROR
-        if let Some(_x) = y {
-            let _v = 1;
-        } else {
             break
```
Comment (Wafflespeanut) [inline tests/compile-fail/while_loop.rs]: a duplicate test.

File: tests/compile-fail/while_loop.rs [hunk 2]
```diff
@@ -30,12 +22,13 @@ fn main() {
...
     }
-    loop { // no error, match is not the only statement
-        match y {
-            Some(_x) => true,
+    loop { //~ERROR
+        let z = match y {
+            Some(x) => x,
             None => break
         };
-        let _x = 1;
```
Comment (Wafflespeanut) [inline tests/compile-fail/while_loop.rs]: this should throw error (even in the now revised version)

File: src/loops.rs [hunk 3]
```diff
@@ -286,7 +294,7 @@ fn is_iterable_array(ty: ty::Ty) -> bool {
 /// If block consists of a single expression (with or without semicolon), return it.
 fn extract_single_expr(block: &Block) -> Option<&Expr> {
     match (&block.stmts.len(), &block.expr) {
-        (&1, &None) => match block.stmts[0].node {
+        (_, &None) => match block.stmts[0].node {
```
Comment (llogiq) [inline src/loops.rs]: If you ignore it, there's no need to get the arg length in the first place.

Comment (Wafflespeanut) [inline src/loops.rs]: alright, yeah, didn't notice that :)

File: tests/compile-fail/while_loop.rs (hunk 2 above)
Comment (llogiq) [inline tests/compile-fail/while_loop.rs]: Why should it throw error? Is the error correct or do we want to do something about it?

Comment (Wafflespeanut) [inline tests/compile-fail/while_loop.rs]: @llogiq that could still be `while...let`, as told in #262 which is the inspiration for this PR :)

**Comment by Wafflespeanut** [comment]
I'd also thought of printing all the statements within the block while suggesting the `while...let`. Though it does that for now, it'll be ugly for large blocks. I'm not sure, shall we go for ellipsis instead? If so, how much should be truncated (everything other than the first line would be my guess)

File: src/loops.rs [hunk 4]
```diff
@@ -161,8 +162,21 @@ impl LateLintPass for LoopsPass {
...
+            let inner_expr = extract_single_expr(block);
+            let extracted = match inner_stmt {
+                Some(_) => inner_stmt,
+                None => inner_expr,
+            };
+
+            if let Some(inner) = extracted {
+                // collect remaining expressions below the match
+                let other_stuff = (1..block.stmts.len())
+                                  .map(|i| {
```
Comment (Manishearth) [inline src/loops.rs]: `block.stmts.iter().map`

Another use case for our iterator lint :p

Comment (Wafflespeanut) [inline src/loops.rs]: I swear I'd written that for comparing something inside the `map`, but I don't remember it now *(sigh)*

Comment (llogiq) [inline src/loops.rs]: You're in good company. When we started using clippy on servo and rustc, we got a lot of warnings, many of them legit.

File: src/loops.rs [hunk 5]
```diff
@@ -161,8 +162,23 @@ impl LateLintPass for LoopsPass {
         // check for `loop { if let {} else break }` that could be `while let`
```
Comment (Manishearth) [inline src/loops.rs]: this comment should be updated

File: src/loops.rs [hunk 6]
```diff
@@ -283,16 +306,31 @@ fn is_iterable_array(ty: ty::Ty) -> bool {
     }
 }
 
+/// If a statement (possibly a `let` binding) consists of an expression, return it.
```
Comment (Manishearth) [inline src/loops.rs]: misnomer, this extracts from a block

File: src/loops.rs [hunk 7]
```diff
@@ -283,16 +306,31 @@ fn is_iterable_array(ty: ty::Ty) -> bool {
...
+fn extract_single_expr_from_stmt(block: &Block) -> Option<&Expr> {
+    match block.expr {
+        Some(_) => None,
+        None => match block.stmts[0].node {
```
Comment (Manishearth) [inline src/loops.rs]: what about blocks with multiple statements?

File: src/loops.rs [hunk 8]
```diff
@@ -283,16 +306,31 @@ fn is_iterable_array(ty: ty::Ty) -> bool {
...
+
 /// If block consists of a single expression (with or without semicolon), return it.
 fn extract_single_expr(block: &Block) -> Option<&Expr> {
-    match (&block.stmts.len(), &block.expr) {
-        (&1, &None) => match block.stmts[0].node {
-            StmtExpr(ref expr, _) |
-            StmtSemi(ref expr, _) => Some(expr),
+    match block.expr {
+        Some(ref expr) => Some(expr),
+        None => match block.stmts[0].node {
```
Comment (Manishearth) [inline src/loops.rs]: What if a block has multiple statements? The comments and function name should be updated

File: src/utils.rs [hunk 9]
```diff
@@ -130,12 +130,17 @@ pub fn snippet_block<'a, T: LintContext>(cx: &T, span: Span, default: &'a str) -
 }
 
 /// Like snippet_block, but add braces if the expr is not an ExprBlock
-pub fn expr_block<'a, T: LintContext>(cx: &T, expr: &Expr, default: &'a str) -> Cow<'a, str> {
+pub fn expr_block<'a, T: LintContext>(cx: &T, expr: &Expr,
```
Comment (Manishearth) [inline src/utils.rs]: I'm not sure if we really need this "option" argument, and if so, it should be better documented

**Comment by Manishearth** [comment]
Mostly LGTM. Needs more comments though.

**Comment by Manishearth** [comment]
Thanks!

**Comment by Wafflespeanut** [comment]
Thank you :)

**Pull request merged**